### PR TITLE
Added unit tests

### PR DESCRIPTION
**Pull request by supunharsha**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java [hunk 1]
```diff
@@ -130,9 +151,30 @@
                 {true, false, true, 0L, 0L, 3600L, 0L, true, TOKEN_STATE_REVOKED, true},
                 {true, false, true, 0L, 0L, 0L, 0L, true, TOKEN_STATE_ACTIVE, true},
                 {true, false, true, 0L, 0L, 0L, 3600L, true, TOKEN_STATE_ACTIVE, true},
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java]: remove this empty line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java [hunk 2]
```diff
@@ -130,9 +151,30 @@
...
+                {true, false, true, 0L, 0L, -1L, 3600L, true, TOKEN_STATE_ACTIVE, true},
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java]: You can use OAuthConstants.TokenStates.ACTIVE instead TOKEN_STATE_ACTIVE. This applies for all other cases

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java [hunk 3]
```diff
@@ -233,6 +273,168 @@ public void testIssue(boolean cacheEnabled,
...
+        return new Object[][] {
+            {oAuthTokenReqMessageContext1, false, false},
+            {oAuthTokenReqMessageContext1, true, false},
+            {oAuthTokenReqMessageContext2, false, true},
+            {oAuthTokenReqMessageContext2, true, true},
+            {oAuthTokenReqMessageContext3, false, false},
+            {oAuthTokenReqMessageContext3, true, false}
+        };
+    }
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java]: remove this empty line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+/**
+ * This class defines unit test for AbstractValidator class
+ */
+public class AbstractValidatorTest extends PowerMockTestCase {
+
+    AbstractValidator abstractValidator;
+    private static final String KEY   = "key";
+    private static final String VALUE = "value";
+    private static final String NOT_ALLOWED_PARAM = "notAllowedParam";
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractValidatorTest.java]: remove this empty line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+import static org.testng.AssertJUnit.assertNotNull;
+import static org.testng.AssertJUnit.assertTrue;
+
+/**
+ * This class defines unit test for AuthorizationCodeGrantHandler class
+ */
+@PrepareForTest({LogFactory.class, OAuthServerConfiguration.class, AuthorizationCodeGrantHandler.class,
+        AppInfoCache.class, OAuth2Util.class, IdentityUtil.class, OAuthCache.class})
+public class AuthorizationCodeGrantHandlerTest extends PowerMockTestCase {
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: remove this empty line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                {oAuthTokenReqMessageContext, authzCodeDO1, true, true, System.currentTimeMillis() + 250000L, false},
+                {oAuthTokenReqMessageContext, authzCodeDO2, true, true, System.currentTimeMillis() + 250000L, false},
+                {oAuthTokenReqMessageContext, authzCodeDO2, true, false, System.currentTimeMillis() + 250000L, false},
+                {oAuthTokenReqMessageContext, authzCodeDO3, true, false, System.currentTimeMillis() + 250000L, true},
+                {oAuthTokenReqMessageContext, authzCodeDO3, true, true, System.currentTimeMillis() + 250000L, true},
+                {oAuthTokenReqMessageContext, authzCodeDO3, true, false, 1000L, false},
+                {oAuthTokenReqMessageContext, authzCodeDO3, true, true, 1000L, false},
+                {oAuthTokenReqMessageContext, authzCodeDO3, false, true, 1000L, false}
+        };
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: remove this empty line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        mockStatic(OAuthServerConfiguration.class);
+        oAuthServerConfiguration = mock(OAuthServerConfiguration.class);
+        when(OAuthServerConfiguration.getInstance()).thenReturn(oAuthServerConfiguration);
+        when(oAuthServerConfiguration.getValueForIsRefreshTokenAllowed(OAuthConstants.GrantTypes.AUTHORIZATION_CODE)).
+                thenReturn(true, false);
+
+        assertTrue(authorizationCodeGrantHandler.issueRefreshToken());
+
+        assertFalse(authorizationCodeGrantHandler.issueRefreshToken());
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: remove this empty line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: remove this empty line

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java (hunk 1 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java]: done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java (hunk 3 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandlerTest.java]: done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractValidatorTest.java (hunk 4 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractValidatorTest.java]: done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java (hunk 5 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java (hunk 6 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java (hunk 7 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java (hunk 8 above)
Comment (supunharsha) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AuthorizationCodeGrantHandlerTest.java]: done

**Comment by mefarazath** [comment]
@supunharsha shall we fix merge conflicts to get this PR merged?

**Pull request merged**